File: common/src/main/java/wile/redstonepen/blocks/track/TrackNet.java
/*
 * @file TrackNet.java
 * @author Stefan Wilhelm (wile)
 * @copyright (C) 2020 Stefan Wilhelm
 * @license MIT (see https://opensource.org/licenses/MIT)
 */
package wile.redstonepen.blocks.track;

import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;

import java.util.List;

public class TrackNet
{
  final List<BlockPos> neighbour_positions;
  final List<Direction> neighbour_sides;
  final List<Direction> internal_sides;
  final List<Direction> power_sides;
  private int power;

  public TrackNet(List<BlockPos> positions, List<Direction> ext_sides, List<Direction> int_sides, List<Direction> pwr_sides)
  { neighbour_positions=positions; neighbour_sides=ext_sides; internal_sides=int_sides; power_sides=pwr_sides; power=0; }

  public TrackNet(List<BlockPos> positions, List<Direction> ext_sides, List<Direction> int_sides, List<Direction> pwr_sides, int power_setval)
  { neighbour_positions=positions; neighbour_sides=ext_sides; internal_sides=int_sides; power_sides=pwr_sides; power=power_setval; }

  int getPower() { return power; }
  void setPower(int p) { power = p; }

  @Override
  public String toString() {
    String s = "NET{";
    s += "p:" + power;
    s += ", intsides:" +  String.join("", internal_sides.stream().map(TrackBlockEntity::dirstr).toList());
    s += ", pwrsides:" +  String.join("", power_sides.stream().map(TrackBlockEntity::dirstr).toList());
    s += ", nbsides:" +  String.join("", neighbour_sides.stream().map(TrackBlockEntity::dirstr).toList());
    s += ", nbpos:" +  String.join(",", neighbour_positions.stream().map(TrackBlockEntity::posstr).toList());
    return s + "}";
  }
}

File: common/src/main/kotlin/wile/redstonepen/blocks/track/TrackNet.kt
package wile.redstonepen.blocks.track

import net.minecraft.core.BlockPos
import net.minecraft.core.Direction

class TrackNet(
    val neighbourPositions: List<BlockPos>,
    val neighbourSides: List<Direction>,
    val internalSides: List<Direction>,
    val powerSides: List<Direction>,
    internal var power: Int = 0,
) {

    override fun toString(): String {
        var s = "NET{"
        s += "p:$power"
        s += ", intsides:" + internalSides.joinToString("") { TrackBlockEntity.dirstr(it) }
        s += ", pwrsides:" + powerSides.joinToString("") { TrackBlockEntity.dirstr(it) }
        s += ", nbsides:" + neighbourSides.joinToString("") { TrackBlockEntity.dirstr(it) }
        s += ", nbpos:" + neighbourPositions.joinToString(",") { TrackBlockEntity.posstr(it) }
        return "$s}"
    }
}
